### ISSUE TITLE
Problem upgrading / running migrations from v4.0.15 to v4.1.2 in Docker

### ISSUE THREAD

**Issue by adam-dej** [snipe/snipe-it#4376]
#### Expected Behavior (or desired behavior if a feature request)

Migrations run successfully after running `php artisan migrate` and application works afterwards.

-----

#### Actual Behavior

Running `php artisan migrate` produces the following error message:

```
                                                                                     
  [BadMethodCallException]                                                           
  Call to undefined method Illuminate\Database\Query\Builder::addEagerConstraints()  
                                                                                     
```

Stacktrace from the log:

```
[2017-11-03 10:02:46] production.ERROR: BadMethodCallException: Call to undefined method Illuminate\Database\Query\Builder::addEagerConstraints() in /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:2450
Stack trace:
#0 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1254): Illuminate\Database\Query\Builder->__call('addEagerConstra...', Array)
#1 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1360): Illuminate\Database\Eloquent\Builder->__call('addEagerConstra...', Array)
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(501): Illuminate\Database\Eloquent\Model->__call('addEagerConstra...', Array)
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(479): Illuminate\Database\Eloquent\Builder->eagerLoadRelation(Array, 'assetloc', Object(Closure))
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(447): Illuminate\Database\Eloquent\Builder->eagerLoadRelations(Array)
#5 /var/www/html/database/migrations/2017_10_27_192423_migrate_denormed_asset_locations.php(58): Illuminate\Database\Eloquent\Builder->get()
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(357): MigrateDenormedAssetLocations->up()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(363): Illuminate\Database\Migrations\Migrator->Illuminate\Database\Migrations\{closure}()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(177): Illuminate\Database\Migrations\Migrator->runMigration(Object(MigrateDenormedAssetLocations), 'up')
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(146): Illuminate\Database\Migrations\Migrator->runUp('/var/www/html/d...', 5, false)
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(95): Illuminate\Database\Migrations\Migrator->runPending(Array, Array)
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Console/Migrations/MigrateCommand.php(69): Illuminate\Database\Migrations\Migrator->run(Array, Array)
#12 [internal function]: Illuminate\Database\Console\Migrations\MigrateCommand->fire()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(29): call_user_func_array(Array, Array)
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(87): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(31): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(539): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(182): Illuminate\Container\Container->call(Array)
#18 /var/www/html/vendor/symfony/console/Command/Command.php(262): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(167): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#20 /var/www/html/vendor/symfony/console/Application.php(888): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#21 /var/www/html/vendor/symfony/console/Application.php(224): Symfony\Component\Console\Application->doRunCommand(Object(Illuminate\Database\Console\Migrations\MigrateCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#22 /var/www/html/vendor/symfony/console/Application.php(125): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(122): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#24 /var/www/html/artisan(35): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#25 {main}  
[2017-11-03 10:02:46] production.ERROR: BadMethodCallException: Call to undefined method Illuminate\Database\Query\Builder::addEagerConstraints() in /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:2450
Stack trace:
#0 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1254): Illuminate\Database\Query\Builder->__call('addEagerConstra...', Array)
#1 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1360): Illuminate\Database\Eloquent\Builder->__call('addEagerConstra...', Array)
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(501): Illuminate\Database\Eloquent\Model->__call('addEagerConstra...', Array)
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(479): Illuminate\Database\Eloquent\Builder->eagerLoadRelation(Array, 'assetloc', Object(Closure))
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(447): Illuminate\Database\Eloquent\Builder->eagerLoadRelations(Array)
#5 /var/www/html/database/migrations/2017_10_27_192423_migrate_denormed_asset_locations.php(58): Illuminate\Database\Eloquent\Builder->get()
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(357): MigrateDenormedAssetLocations->up()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(363): Illuminate\Database\Migrations\Migrator->Illuminate\Database\Migrations\{closure}()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(177): Illuminate\Database\Migrations\Migrator->runMigration(Object(MigrateDenormedAssetLocations), 'up')
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(146): Illuminate\Database\Migrations\Migrator->runUp('/var/www/html/d...', 5, false)
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(95): Illuminate\Database\Migrations\Migrator->runPending(Array, Array)
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Console/Migrations/MigrateCommand.php(69): Illuminate\Database\Migrations\Migrator->run(Array, Array)
#12 [internal function]: Illuminate\Database\Console\Migrations\MigrateCommand->fire()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(29): call_user_func_array(Array, Array)
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(87): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(31): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(539): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(182): Illuminate\Container\Container->call(Array)
#18 /var/www/html/vendor/symfony/console/Command/Command.php(262): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(167): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#20 /var/www/html/vendor/symfony/console/Application.php(888): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#21 /var/www/html/vendor/symfony/console/Application.php(224): Symfony\Component\Console\Application->doRunCommand(Object(Illuminate\Database\Console\Migrations\MigrateCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#22 /var/www/html/vendor/symfony/console/Application.php(125): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(122): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#24 /var/www/html/artisan(35): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#25 {main} 
```

I've also tried to migrate from v4.0.15 to v4.1.0 with the same result.

-----

#### Please confirm you have done the following before posting your bug report:

- [X] I have enabled debug mode 
- [X] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

-----
#### Provide answers to these questions:

- Is this a fresh install or an upgrade? **Upgrade from v4.0.15**
- Version of Snipe-IT you're running **Trying to upgrade to v4.1.2**
- Version of PHP you're running **7.0.22-0ubuntu0.16.04.1, included in the docker image**
- Version of MySQL/MariaDB you're running **mariadb:10.1**
- What OS and web server you're running Snipe-IT on **Docker container running on CoreOS**
- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc) **docker**
- WITH DEBUG TURNED ON, if you're getting an error in your browser, include that error **not applicable**
- What specific Snipe-IT page you're on, and what specific element you're interacting with to trigger the error **not applicable**
- If a stacktrace is provided in the error, include that too. **included**
- Any errors that appear in your browser's error console. **not applicable**
- Confirm whether the error is reproduceable on the demo: https://snipeitapp.com/demo. **not applicable**
- Include any additional information you can find in `app/storage/logs` and your webserver's logs. **not applicable**
- Include what you've done so far in the installation, and if you got any error messages along the way. **this is the first step after pulling container v4.1.2**
- Indicate whether or not you've manually edited any data directly in the database **no**

Please do not post an issue without answering the related questions above. If you have opened a different issue and already answered these questions, answer them again, once for every ticket. It will be next to impossible for us to help you.

https://snipe-it.readme.io/docs/getting-help

**Comment by chibacityblues** [comment]
I can confirm this issue. Same thing happened when I tried to upgrade our installation from v4.0.15 to v4.1.3.

**Comment by snipe** [comment]
Okay, try commenting out the contents of that migration's `up()` method. Run migrations again, and then run `php artisan snipeit:sync-asset-locations --output=all`

**Comment by snipe** [comment]
(That artisan command is only available in 4.1.3)

**Comment by chibacityblues** [comment]
Thanks snipe, tried it. Is there a way to make the command more verbose? This is the output I get:

> 1045/1046 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░]  99%
> 
>   [ErrorException]
>   Trying to get property of non-object

**Comment by snipe** [comment]
Adding the -vvv flag will give you verbose output (or you can tail your error logs)

`php artisan snipeit:sync-asset-locations --output=all -vvv`

**Comment by chibacityblues** [comment]
```

Exception trace:
 () at /var/www/html/app/Console/Commands/SyncAssetLocations.php:100
 Illuminate\Foundation\Bootstrap\HandleExceptions->handleError() at /var/www/html/app/Console/Commands/SyncAssetLocations.php:100
 App\Console\Commands\SyncAssetLocations->handle() at n/a:n/a
 call_user_func_array() at /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:29
 Illuminate\Container\BoundMethod::Illuminate\Container\{closure}() at /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:87
 Illuminate\Container\BoundMethod::callBoundMethod() at /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php:31
 Illuminate\Container\BoundMethod::call() at /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php:539
 Illuminate\Container\Container->call() at /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php:182
 Illuminate\Console\Command->execute() at /var/www/html/vendor/symfony/console/Command/Command.php:262
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php:167
 Illuminate\Console\Command->run() at /var/www/html/vendor/symfony/console/Application.php:888
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/console/Application.php:224
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/symfony/console/Application.php:125
 Symfony\Component\Console\Application->run() at /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php:122
 Illuminate\Foundation\Console\Kernel->handle() at /var/www/html/artisan:35
```

**Comment by snipe** [comment]
Wow, that's not a remotely helpful stacktrace. Sigh. 

If you tail your `storage/logs/laravel.log` while you run that command, does it give you anything more useful?

**Comment by chibacityblues** [comment]
This looks a bit more useful: 

```
[2017-11-04 23:00:29] production.ERROR: ErrorException: Trying to get property of non-object in /var/www/html/app/Console/Commands/SyncAssetLocations.php:100
Stack trace:
#0 /var/www/html/app/Console/Commands/SyncAssetLocations.php(100): Illuminate\Foundation\Bootstrap\HandleExceptions->handleError(8, 'Trying to get p...', '/var/www/html/a...', 100, Array)
#1 [internal function]: App\Console\Commands\SyncAssetLocations->handle()
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(29): call_user_func_array(Array, Array)
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(87): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(31): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(539): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(182): Illuminate\Container\Container->call(Array)
#7 /var/www/html/vendor/symfony/console/Command/Command.php(262): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(167): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#9 /var/www/html/vendor/symfony/console/Application.php(888): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#10 /var/www/html/vendor/symfony/console/Application.php(224): Symfony\Component\Console\Application->doRunCommand(Object(App\Console\Commands\SyncAssetLocations), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#11 /var/www/html/vendor/symfony/console/Application.php(125): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#12 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(122): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#13 /var/www/html/artisan(35): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 {main}
```

**Comment by snipe** [comment]
Do you have any deleted assets or locations?

**Comment by snipe** [comment]
If you can pull the latest from develop, I just pushed out a fix to the sync script. Give that a try. I think that error is happening when something in the relationship is broken, for example if an asset (even a deleted one) is assigned to a location that got deleted, etc.

**Comment by chibacityblues** [comment]
We got quite a lot deleted assets from some borked imports back in the 2.x days. As a quick test, I tried deleting every asset with a not null deleted_at - doesn't change anything. I'll revert my database and try your fix asap.

**Comment by snipe** [comment]
Sorry about this. There are so many different states the data can be in, and while we tried to consider all of the normal ones, folks who have been with us longest are most likely to have some weird data, since we didn't always check data integrity as vigorously as we do now. (For example, not allowing you to delete a location that has associated assets, people, etc.) We sometimes forget that the data won't always be the way it would be if you started using Snipe-IT today for the first time.

So basically your reward for being a long-time-user is complete fuckery. Sorry. :(

**Comment by adam-dej** [comment]
I've tried to run migrations again, using `develop` image. This time the error is different:

`Exception: Asset assignment Loop for Asset ID: 17`

That asset was checked-out to another asset, which however was not checked out and was sitting
at its default location. After I've used an older image to check-in that asset, the error repeated
for asset ID 345.

```
[2017-11-05 15:58:53] production.ERROR: Exception: Asset assignment Loop for Asset ID: 345 in /var/www/html/app/Models/Asset.php:270
Stack trace:
#0 /var/www/html/app/Console/Commands/SyncAssetLocations.php(108): App\Models\Asset->assetLoc()
#1 [internal function]: App\Console\Commands\SyncAssetLocations->handle()
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(29): call_user_func_array(Array, Array)
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(87): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(31): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(539): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(182): Illuminate\Container\Container->call(Array)
#7 /var/www/html/vendor/symfony/console/Command/Command.php(262): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArrayInput), Object(Illuminate\Console\OutputStyle))
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(167): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArrayInput), Object(Illuminate\Console\OutputStyle))
#9 /var/www/html/vendor/symfony/console/Application.php(888): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArrayInput), Object(Symfony\Component\Console\Output\BufferedOutput))
#10 /var/www/html/vendor/symfony/console/Application.php(224): Symfony\Component\Console\Application->doRunCommand(Object(App\Console\Commands\SyncAssetLocations), Object(Symfony\Component\Console\Input\ArrayInput), Object(Symfony\Component\Console\Output\BufferedOutput))
#11 /var/www/html/vendor/symfony/console/Application.php(125): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArrayInput), Object(Symfony\Component\Console\Output\BufferedOutput))
#12 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Application.php(141): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArrayInput), Object(Symfony\Component\Console\Output\BufferedOutput))
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(220): Illuminate\Console\Application->call('snipeit:sync-as...', Object(Illuminate\Support\Collection), NULL)
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(221): Illuminate\Foundation\Console\Kernel->call('snipeit:sync-as...', Array)
#15 /var/www/html/database/migrations/2017_10_27_192423_migrate_denormed_asset_locations.php(16): Illuminate\Support\Facades\Facade::__callStatic('call', Array)
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(357): MigrateDenormedAssetLocations->up()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(363): Illuminate\Database\Migrations\Migrator->Illuminate\Database\Migrations\{closure}()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(177): Illuminate\Database\Migrations\Migrator->runMigration(Object(MigrateDenormedAssetLocations), 'up')
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(146): Illuminate\Database\Migrations\Migrator->runUp('/var/www/html/d...', 5, false)
#20 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Migrations/Migrator.php(95): Illuminate\Database\Migrations\Migrator->runPending(Array, Array)
#21 /var/www/html/vendor/laravel/framework/src/Illuminate/Database/Console/Migrations/MigrateCommand.php(69): Illuminate\Database\Migrations\Migrator->run(Array, Array)
#22 [internal function]: Illuminate\Database\Console\Migrations\MigrateCommand->fire()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(29): call_user_func_array(Array, Array)
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(87): Illuminate\Container\BoundMethod::Illuminate\Container\{closure}()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(31): Illuminate\Container\BoundMethod::callBoundMethod(Object(Illuminate\Foundation\Application), Array, Object(Closure))
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(539): Illuminate\Container\BoundMethod::call(Object(Illuminate\Foundation\Application), Array, Array, NULL)
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(182): Illuminate\Container\Container->call(Array)
#28 /var/www/html/vendor/symfony/console/Command/Command.php(262): Illuminate\Console\Command->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(167): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Illuminate\Console\OutputStyle))
#30 /var/www/html/vendor/symfony/console/Application.php(888): Illuminate\Console\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#31 /var/www/html/vendor/symfony/console/Application.php(224): Symfony\Component\Console\Application->doRunCommand(Object(Illuminate\Database\Console\Migrations\MigrateCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#32 /var/www/html/vendor/symfony/console/Application.php(125): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(122): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#34 /var/www/html/artisan(35): Illuminate\Foundation\Console\Kernel->handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#35 {main} 
```

That happened for several more assets, usually deployed to other assets (in this case, the second asset is deployed to a location):

```
MariaDB> select assigned_to,assigned_type from assets where id=388;
+-------------+------------------+
| assigned_to | assigned_type    |
+-------------+------------------+
|         382 | App\Models\Asset |
+-------------+------------------+
1 row in set (0.00 sec)

MariaDB> select assigned_to,assigned_type from assets where id=382;
+-------------+---------------------+
| assigned_to | assigned_type       |
+-------------+---------------------+
|          42 | App\Models\Location |
+-------------+---------------------+
1 row in set (0.00 sec)
```

I've tried to comment-out body of the migration and run
`php artisan snipeit:sync-asset-locations --output=all -vvv`, leading to the same result.

After the error repeated several times, I have stopped to check-in assets using an older image,
but I've instead (for debugging purposes only of course) edited `assigned_to` field of
affected assets to `NULL`. After editing about 50 assets the command run successfully, output
ending with `Assets still without a location: 50`, and a list of assets most of which I have
checked in. Application then appeared to work.

Additional info: as opposed to chibacityblues, I'm not a long-time user, I've deployed Snipe-IT
about two weeks ago, I have ~477 assets (and 2 deleted), ~50 locations (none deleted). I've
utilized the API heavily to automate certain tasks related to system adoption (mainly mass-setting
default locations of assets using a barcode reader).

If there is no other way I'll write a script that will check-in affected assets, update, migrate,
and check them out again, but I'd really prefer finding (or helping you find) a root cause of
this problem.

**Comment by adam-dej** [comment]
I've managed to migrate successfully using `develop` image.

After editing function `assetLoc()` in `app/Models/Asset.php` to:

```php
public function assetLoc()
{
    $seenAssets = [];
    return $this->recursivelyGetAssetLoc($seenAssets);
}

public function recursivelyGetAssetLoc($seenAssets)
{
    static $first_asset;
    if (!empty($this->assignedType())) {
        if ($this->assignedType() == self::ASSET) {
            if(!$first_asset) {
                $first_asset=$this;
            }
            if (in_array($this->id, $seenAssets)) { 
                throw new \Exception("Asset assignment Loop for Asset ID: ".$first_asset->id);
            }
            $seenAssets[] = $this->id;
            $assigned_to=Asset::find($this->assigned_to); //have to do this this way because otherwise it errors
            if ($assigned_to) {
                return $assigned_to->recursivelyGetAssetLoc($seenAssets);
            } // Recurse until we have a final location
        }
        if ($this->assignedType() == self::LOCATION) {
            if ($this->assignedTo) {
                return $this->assignedTo;
            }

        }
        if ($this->assignedType() == self::USER) {
            if (($this->assignedTo) && $this->assignedTo->userLoc) {
                return $this->assignedTo->userLoc;
            }
            //this makes no sense
            return $this->defaultLoc;

        }

    }
    return $this->defaultLoc;
}
```

migrations work and the app is running afterwards. I'd submit this as a PR,  however using an array to check if an asset was already seen is inefficient, and using `\Ds\Set` would require adding a dependency (I think, I don't use PHP watsoever).

**Comment by snipe** [comment]
You should definitely not edit assetLoc as that will absolutely break things elsewhere in the system.

We already patched develop and master to remove that migration and turned it into an artisan command instead.

`Asset assignment Loop for Asset ID: 17`

That means an asset is assigned to itself. At the very least, it means something in the chain of assignment is broken/wrong. It's a data issue.

You should correct that, as it can cause really strange behavior in the system.
